doc/decisions/simplify_api.md

```diff
@@ -0,0 +1,94 @@
+# Simplify API
+
+## Problem
+
+According to src/libs/elektra/symbols.map the public core API has 124 functions, which
+is arguable too much for a key-value API which has as highest goal simplicity.
+
+In particular following areas have many functions and are not simple:
+
+- binary keys
+- memory functions
+- comparison functions
+- error/warnings
+
+The urgency of this decision is that API can be easily introduced later but we cannot
+get rid of it after 1.0.
+
+## Constraints
+
+To not disturb main features such as:
+
+-
+
+## Assumptions
+
+Binary data is not a core feature, if needed the plugin system can also work without (properly tagged) binary data in key sets.
+
+## Considered Alternatives
+
+- flags for binary data
+
+## Decision
+
+Remove:
+
+- keyGetMeta (@kodebach)
+- keySetMeta (@kodebach)
+- keyRewindMeta
+- keyNextMeta
+- keyCurrentMeta
+- keyCopyAllMeta
+- keyCopyMeta
+- keyGetBaseName
+- keyGetBaseNameSize
+- keyGetBinary (\*buffers.md)
+- keyGetName
+- keyGetNameSize
+- ksHead
+- ksTail
+- ksCopyInternal
+- keyClear (clarification with keyCopy needed)
+- keyCompare (done)
+- keyCompareMeta (done)
+
+Make private:
+
+- elektraStrCaseCmp;
+- elektraStrCmp;
+- elektraStrDup;
+- elektraStrLen;
+- elektraStrNCaseCmp;
+- elektraStrNCmp;
+- elektraVFormat;
+- ksDeepDup
+- ksGetAlloc
+- ksInit
+- keyGetRef
+
+Unclear:
+
+- keyCmp
+- ksPop
+- keyNeedSync
+- keyIsBelow
+- keyIsBelowOrSame
+- keyIsDirectlyBelow
+- keyIsBinary (binary issue)
+- keyIsString (binary issue)
+- keyName (
+- keyGetBaseName
+- keyGetString
+- keySetString
+- keyGetBinary
+- keySetBinary
+- ksClear
+- ksCut
```

Comment:
To keep @lawli3t and your work in the defined scope we must, as far as possible, only talk about reduction of API.

I think all of the mentioned functions can be implemented outside of core, so simply removing ksCut should be the obvious choice?

Btw. there is also the idea of ksForEach and ksFilter, see tests/abi/testabi_ks.c which also let you easily implement ksCut.

Reply:
> I think all of the mentioned functions can be implemented outside of core

In theory yes, but this would a bad idea.

- `ksFindHierarchy`: Since we know have `ksSearch`, this would be the most feasible to move out of the core. However, the optimizations the avoid allocating a new `Key` when searching for `end` are only possible in the core: https://github.com/ElektraInitiative/libelektra/blob/f0b07d837f7e24e55e4a1cbb65f83ea4ac68a151/src/libs/elektra/keyset.c#L1246-L1270
- `ksCopyRange`: If this was done outside the core, you need to do a separate `memcpy` for every `Key *`. Inside the core, we can just use a single `memcpy` to copy the whole section of the `KeySet`. 
- `ksRemoveRange`: Here the multiple `memcpy` issue is even worse, because we need to always "close the gap" in the array of the `KeySet`. For example, if we have a keyset with 1000 Keys and want to remove the first 900 of them, this would result in 900 `memmove` moving 900, then 899, 898, ... Keys at a time. The optimized version doesn't need a memmove at all, since we are removing until the end, but even if we kept the last key, it would be a single `memmove` of 899 keys instead of 899 separate calls.

> Btw. there is also the idea of ksForEach and ksFilter, see tests/abi/testabi_ks.c which also let you easily implement ksCut.

`ksForEach` is pretty useless now that we can simple use an indexed `for` loop. `ksFilter` could still be useful, but has several problems:
- It uses `keyDup`, so the behaviour is different.
- C doesn't support lambdas so using local data is hard in general, and impossible in this case, because the don't have an additional `void *` argument.
- It has the same multiple `memcpy` instead of a single one problem that is mentioned above.
- It cannot be used with the optimized search for the end key, that `ksFindHierarchy` and `ksCut` use. You would have to call `ksFindHierarchy` first and then pass the result to the filter, but is not possible because of the lambda issue.

> To keep @lawli3t and your work in the defined scope we must, as far as possible, only talk about reduction of API.

I don't think this change would actually be that much work (probably comparable to removing `ksCut`).

I'm not sure what the best way forward is here. I can't really see a reason why `ksCut` should be in the 1.0 API. Compared to the other functions it is pretty specific. However, we can't just remove it, because there is no good replacement.

There is also another opportunity here:
- `ksFindHierarchy` and `ksSearch` could be combined into
   ```c
   /**
    * searches for root and the hierarchy below root in ks
    * if root is in ks:
    *    - sets start to the position of root
    *    - sets end to the position of the first key that is not equal to or below root
    *    - returns true
    * if root is not in ks, but ks contains a key that is below root:
    *    - sets start to the position where root would be inserted, i.e. the first key inside ks that is sorted after root
    *    - sets end to the same value as start
    *    - returns false
    * if root is not in ks, and ks contains no key that is below root:
    *    - sets start to ksGetSize (ks)
    *    - sets end to ksGetSize (ks)
    *    - returns false
    */
   bool ksFind (const KeySet * ks, const Key * root, elektraCursor * start, elektraCursor * end);

   // shouldn't actually be part of the core API, but a helper for C only
   static inline elektraCursor ksFindHierarchy (const KeySet * ks, const Key * root, elektraCursor * end) {
     elektraCursor start, end;
     bool found = ksFind (ks, root, &start, &end);
     if (found) {
       return start;
     }
     if (end != NULL) {
       *end = ksGetSize (ks);
     }
     return ksGetSize (start);
   }

   // shouldn't actually be part of the core API, just here to show the replacement
   static inline ssize_t ksSearch (const KeySet * ks, const Key * key) {
     elektraCursor start;
     bool found = ksFind (ks, root, &start, NULL);
     return found ? start : -start - 1;
   }
   ```
   In fact, if properly optimized, this `ksFind` could be used for `ksLookup` with no options.
- `ksRemoveRange` can be merged into `ksCopyRange`:
  ```c
  // new, copies keys in range start <= index < end from source to dest
  // if remove == true, the copied keys are also removed from source
  // if dest == NULL, no copying happens, i.e. removed keys are deleted
  // use ksCopyRange (ks, start, -1) to copy all keys from start to the end of ks
  // returns the number of copied/removed keys
  ssize_t ksCopyRange (KeySet * dest, KeySet * source, elektraCursor start, elektraCursor end, bool remove);

  static inline ssize_t ksRemoveRange (KeySet * ks, elektraCursor start, elektraCursor end) {
    return ksCopyRange (NULL, ks, start, end, true);
  }
  ```

All in all, however, I think it is probably best to keep things as they are for the time being and maybe make `ksCut` internal (removing is probably not feasible).